coderd/updatecheck/updatecheck.go

```diff
@@ -0,0 +1,229 @@
+//
+// The update check is performed by querying the GitHub API for the
+// latest release of Coder.
+package updatecheck
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"io"
+	"net/http"
+	"time"
+
+	"github.com/google/go-github/v43/github"
+	"golang.org/x/mod/semver"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/coderd/database"
+)
+
+const (
+	// defaultURL defines the URL to check for the latest version of Coder.
+	defaultURL = "https://api.github.com/repos/coder/coder/releases/latest"
+)
+
+// Checker is responsible for periodically checking for updates.
+type Checker struct {
+	ctx        context.Context
+	cancel     context.CancelFunc
+	db         database.Store
+	log        slog.Logger
+	opts       Options
+	firstCheck chan struct{}
+	closed     chan struct{}
+}
+
+// Options set optional parameters for the update check.
+type Options struct {
+	// Client is the HTTP client to use for the update check,
+	// if omitted, http.DefaultClient will be used.
+	Client *http.Client
+	// URL is the URL to check for the latest version of Coder,
+	// if omitted, the default URL will be used.
+	URL string
+	// Interval is the interval at which to check for updates,
+	// default 24h.
+	Interval time.Duration
+	// Notify is called when a newer version of Coder (than the
+	// last update check) is available.
+	Notify func(r Result)
+}
+
+// New returns a new Checker that periodically checks for Coder updates.
+func New(db database.Store, log slog.Logger, opts Options) *Checker {
+	if opts.Client == nil {
+		opts.Client = http.DefaultClient
+	}
+	if opts.URL == "" {
+		opts.URL = defaultURL
+	}
+	if opts.Interval == 0 {
+		opts.Interval = 24 * time.Hour
+	}
+	if opts.Notify == nil {
+		opts.Notify = func(r Result) {}
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	c := &Checker{
+		ctx:        ctx,
+		cancel:     cancel,
+		db:         db,
+		log:        log,
+		opts:       opts,
+		firstCheck: make(chan struct{}),
+		closed:     make(chan struct{}),
+	}
+	go c.start()
+	return c
+}
+
+// Result is the result from the last update check.
+type Result struct {
+	Checked time.Time `json:"checked,omitempty"`
+	Version string    `json:"version,omitempty"`
+	URL     string    `json:"url,omitempty"`
+}
+
+// Latest returns the latest version of Coder.
+func (c *Checker) Latest(ctx context.Context) (r Result, err error) {
+	select {
+	case <-c.ctx.Done():
+		return r, c.ctx.Err()
+	case <-ctx.Done():
+		return r, ctx.Err()
+	case <-c.firstCheck:
+	}
+
+	return c.lastUpdateCheck(ctx)
+}
+
+func (c *Checker) init() (Result, error) {
+	defer close(c.firstCheck)
+
+	r, err := c.lastUpdateCheck(c.ctx)
+	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
+		return Result{}, xerrors.Errorf("last update check: %w", err)
+	}
+	if r.Checked.IsZero() || time.Since(r.Checked) > c.opts.Interval {
+		r, err = c.update()
+		if err != nil {
+			return Result{}, xerrors.Errorf("update check failed: %w", err)
+		}
+	}
```

Comment:
I think we should avoid doing an update check immediately on startup, this could be delayed by 10 minutes or something to avoid:

- every replica doing database queries immediately on startup to try to do an update check,
- any potential panics in this code causing coder to instantly crash on start

Reply:
We only do the update check if it's been >24h since the last one. And unless all replicas start up at the same time, this shouldn't be a problem, especially if we add the table locking.

Regarding the panics, ultimately we mustn't have any, but if we did, I would say it's better the sooner it happens. Wouldn't want it to happen while Coder is in the middle of provisioning workspaces, etc.

Reply:
> every replica doing database queries immediately on startup to try to do an update check,

> And unless all replicas start up at the same time, this shouldn't be a problem

It's a common pattern to define a startup/access delay for workers that hit the same resource, usually 0-60 sec. Maybe it can help also here?

Reply:
I'll think about this for a follow up PR, I don't imagine a single database query will be problematic at startup, though. If we want to spread out service starts, it would be good to have a more structured plan than doing it individually and uniquely for each service.